pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+	// and serve HTTP requests received on addr/pattern
+	// if the "pattern" is empty (i.e., ""), no endpoint will be started. An error
+	// will be returned if there is any.
+	StartServing(pattern, addr string, logger promhttp.Logger) error
+
+	// OperationStart takes in an operation and cache it's start time.
+	// if the operation already exists, it's an no-op.
+	OperationStart(op Operation)
+
+	// RecordMetrics records a metric point. Note that it will be an no-op if an
+	// operation has not been marked "Started" previously via invoking "OperationStart".
+	// op - the Operation which the metric is associated with.
+	// state - metrics manager operations differently to the cache based on the state
+	//         of the Operation:
+	//   InterimFailure: record a failure metric however keep operation's start time in cache
+	//   TerminatingFailure: record a failure metric and remove operation's start time from cache
+	//   Success: record a success metric and remove the operation's start time from cache
+	RecordMetrics(op Operation, state OperationState)
+}
+
+// Operation is a structure which holds information to identifier a snapshot
+// related operation
+type Operation struct {
+	// the name of the operation, for example: "CreateSnapshot", "DeleteSnapshot"
+	Name string
+	// the name of the driver which executes the operation
+	Driver string
+	// the resource UID to which the operation has been executed against
+	ResourceID types.UID
+}
+
+func (op *Operation) key() string {
+	return fmt.Sprintf("%s-%s-%s", op.Name, op.Driver, op.ResourceID)
+}
+
+type operationTs struct {
+	operation Operation
+	startTime time.Time
+}
+
+type operationMetricsManager struct {
+	// cache is a concurrent-safe map which stores start timestamps for all
+	// ongoing operations.
+	// key is
+	// value is
+	cache sync.Map
+
+	registry k8smetrics.KubeRegistry
+
+	opLatencyMetric *k8smetrics.HistogramVec
+}
+
+func NewMetricsManager() MetricsManager {
+	mgr := &operationMetricsManager{
+		cache: sync.Map{},
+	}
+	mgr.init()
+	return mgr
+}
+
+func (opMgr *operationMetricsManager) OperationStart(op Operation) {
+	opTs := operationTs{
+		operation: op,
+		startTime: time.Now(),
+	}
+	opMgr.cache.LoadOrStore(op.key(), &opTs)
+}
+
+func (opMgr *operationMetricsManager) RecordMetrics(op Operation, state OperationState) {
+	key := op.key()
+	obj, exists := opMgr.cache.Load(key)
+	if !exists {
+		// the operation has not been cached, return directly
+		return
+	}
+	ts, ok := obj.(*operationTs)
+	if !ok {
+		// the cached item is not an operationTs, should NEVER happen, clean and return
+		klog.Errorf("Invalid cache entry for key %s", key)
+		opMgr.cache.Delete(key)
+		return
+	}
+	duration := time.Since(ts.startTime).Seconds()
+	opMgr.opLatencyMetric.WithLabelValues(op.Driver, op.Name, string(state)).Observe(duration)
+	switch state {
+	case Success, TerminatingFailure:
+		opMgr.cache.Delete(key)
+	case InterimFailure:
+		// do nothing
```

Comment:
I don't think we should count retry loops in the same metric for total latency across retry loops. It's going to make the error rate calculations strange because the total count is number of retry loops instead of number of objects processed.

Reply:
If error rate is the only concern, I think it's a solvable problem, it depends on how you define error rates.
if error rate is defined by total permanent failure / total requests, it can be calculated by get the sum with "TerminatingFailure" over the total number of requests.
if error rate is defined by total failures / total # operations, it can be calculated by sum(terminating failure + interim failure) / total number of operations.
count retrying loops have its value to capture interim status and does make sense to me. WDYT?

Reply:
How do you calculate # total requests vs # total operations? Currently this metric is counting # total operations.

I agree counting retry loops and counting # user requests are both valuable. I think it is confusing and difficult to count them both in one metric.

Reply:
total # of operation is going to be "snapshot_controller_operation_total_seconds_count"
total # of request is going to be(approximately) "snapshot_controller_operation_total_seconds_count" however with labels == success || labels == Permanent failure filtering?

Reply:
The start time doesn't make sense for the single-retry loop case. Say the operation starts at time X. It fails 5 times, and then succeeds on the 6th. Say every loop takes 1 second.

The latency of the temporary failure is going to be recorded as X+1, X+2, X+3, X+4, X+5, instead of 1, 1, 1, 1, 1.

I really think it's easier to have two different metrics: one for e2e time, and one for a single reconciliation loop.

Reply:
that's a very fair point.
I thought about that but IMO it might not necessary a problem as latency metric for an interim reconcile loop(failed one) is really not that important. When error happens, count is more important than how much time it has spent? WDYT?
if you feel strong about latency of a failed reconcile loop is valuable, another timestamp can be introduced to record the last visited time in the cache as well to have a relatively more accurate metrics. I am hesitate to introduce two similar metrics.

Reply:
I think latency of a single loop is valuable to show if something is hanging/slow that causes back pressure on the controller. I'm not sure I understand the confusion around having two metrics, one to capture total e2e time, and one to capture a single reconciliation loop time. In fact, I think the workqueue struct already has generic metrics that we may be able to reuse.

Reply:
workqueue does have generic metrics support at [here](https://github.com/kubernetes/client-go/blob/master/util/workqueue/metrics.go), however it's not trivial to distinguish different operations and  add labels dynamically based on the type of operation is not easy. A customized [MetricsProvider](https://github.com/kubernetes/client-go/blob/master/util/workqueue/metrics.go#L173) is needed.
When I say two metrics, I meant two histograms, we can record use the same histogram as e2e time with some enhancement to the cache.

Reply:
Sorry I still have concerns that that combining the two different measurements into one histogram for the following reasons:
1) Total latency, and per-loop latency are intended for two different audiences. One is for end users, another is for developers. Combining the two measurements into one I think is confusing to explain to end users how to interpret. It requires that the end users understand how things are implemented.
2) I'm not convinced we're able to capture all the conditions we want. For example, can we tell the difference between success latency of a retry loop vs success latency of a total operation? We can potentially keep on adding more states for each of these corner cases, but I think that makes 1) worse.
3) Adding more special cases to the cache makes the cache more complicated, more specialized, and less generalizable to future components.

I still prefer that we limit this to only capture end-to-end latency for now and consider how to capture a per-loop metric separately.

Reply:
removed per operation latency from the current implementation and only left end to end + error count.